### PR TITLE
fix: Clean up ip_address defaulting.

### PR DESCRIPTION
**Pull request by alex-hofsteede**
Make ip_address defaulting happen consistently in one place
and update the tests.

fix: Deal with bad message interface data.

Add a cast to deal with the Message interface being sent as a string.
Ideally this would actuallly just be in the schema and we would reject it,
but we have existing logic to insert 'message' into
'sentry.interface.Message' that I don't want to break.

fixes: SENTRY-58P, SENTRY-58M

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: tests/sentry/coreapi/tests.py [hunk 1]
```diff
@@ -33,9 +33,9 @@ def setUp(self):
         self.pk = self.project.key_set.get_or_create()[0]
         self.helper = self.helper_cls(agent='Awesome Browser', ip_address='198.51.100.0')
 
-    def validate_and_normalize(self, data):
+    def validate_and_normalize(self, data, request_env={}):
```
Comment (mattrobenolt) [inline tests/sentry/coreapi/tests.py]: I know this is a test, but be careful with this `{}`. It'd be saver to use `None`.

File: src/sentry/coreapi.py [hunk 2]
```diff
@@ -561,30 +543,17 @@ def _decode(self):
...
-
-        # we always fill in the IP so that filters and other items can
-        # access it (even if it eventually gets scrubbed)
-        helper.ensure_has_ip(
-            data,
-            self._client_ip,
-            set_if_missing=auth.is_public or
-            data.get('platform') in ('javascript', 'cocoa', 'objc')
-        )
+        data.setdefault('sdk', helper.parse_client_as_sdk(auth.client))
```
Comment (mattrobenolt) [inline src/sentry/coreapi.py]: The only opinion I have here is the rhs is being evaluated regardless if `sdk` is set or not. So even if it's set correctly, we're doing the unnecessary work from `parse_client_as_sdk`.

File: tests/sentry/coreapi/tests.py (hunk 1 above)
Comment (alex-hofsteede) [reply tests/sentry/coreapi/tests.py]: fixed, here and in `normalize()`

File: src/sentry/coreapi.py (hunk 2 above)
Comment (alex-hofsteede) [reply src/sentry/coreapi.py]: fixed

**Comment by alex-hofsteede** [comment]
@mattrobenolt this is ready for a final pass if you have time.

**Review by mattrobenolt — APPROVED** [review]
:cookie:

**Pull request merged**